src/subject_name/verify.rs

```diff
@@ -351,6 +356,39 @@ fn iterate_names(
     }
 }
 
+#[cfg(feature = "alloc")]
+pub(crate) fn list_cert_dns_names<'names>(
+    cert: &'names crate::EndEntityCert<'names>,
+) -> Result<impl Iterator<Item = GeneralDnsNameRef<'names>>, Error> {
+    let cert = &cert.inner();
+    let names = core::cell::RefCell::new(Vec::new());
```

Comment:
Why do we need a `RefCell` here?

Reply:
We need `names` to be mutable from within the handler passed to `iterate_names` in order to `push` onto it. Dropping the `RefCell` and making `mut names` won't work with the existing `iterate_names` helper that takes a `Fn` closure, it produces the err "cannot borrow `names` as mutable, as it is a captured variable in a `Fn` closure". 

I think we could avoid the `RefCell` by changing `iterate_names` to take a `FnMut` and changing the call-sites to pass a `&mut |name| { ... }` `FnMut` closure, or making a new `iterate_names_mut` helper that does that to avoid needing to change unrelated call-sites. 

Do you have a preference between these options? In some ways the `RefCell` makes for a smaller overall diff but it does mean we have runtime borrow checking :-/

Reply:
it looks like `RefCell` is being used because the function that's passed to `iterate_names` is a `Fn` rather than a `FnMut`...but, as far as I can tell, there's no reason we can't just change `iterate_names` to take an `&mut dyn FnMut(GeneralName<'names>) -> NameIteration` rather than an `&dyn Fn(...)`. that might let us clean up some other code in this module as well...

Reply:
>  that might let us clean up some other code in this module as well...

I don't think the other users of `iterate_names` in this module will benefit from the change (_or I'm overlooking what you had in mind!_). To me it seems like `list_cert_dns_names` is the only user of `iterate_names` that manipulates the environment outside of the closure.

Reply:
Yeah, I hadn't looked super closely, but I think you're right that this is the only function that actually needs a mutable closure. But, the other functions will still work if the closure is a `FnMut` instead of a `Fn`.

Reply:
I force pushed a revision that drops the `RefCell` and changes `iterate_names` to use a `FnMut`. Happy to tweak further if @djc has more thoughts :-)

Reply:
Looks great IMO. I'm also a little unsure why the `iterate_names` function takes a reference to a `dyn Fn[Mut]` closure rather than taking an `impl Fn[Mut]`, but it was already like that...

Reply:
>  I'm also a little unsure why the iterate_names function takes a reference to a dyn Fn[Mut] closure rather than taking an impl Fn[Mut]

I figured we might as well fix this up while we're here. I pushed another revision that reworks to use `impl` to avoid the overhead of dynamic dispatch.